Support multiple CIDR addresses for the pod SDN

JacobTanenbaum opened the pull request:
Work in progress PR for multiple CIDR address work. addresses can be defined in the master config file by 


```
networkConfig:
  clusterNetworkCIDR: ""
  clusterNetworkConfig:
  - clusterNetworkCIDR: 10.128.0.0/28
  - clusterNetworkCIDR: 12.128.0.0/28
  externalIPNetworkCIDRs: null

```

or by passing a comma seporated list of them on the command line using --network-cidr="10.128.0.0/28,12.128.0.0/28"


besides general review - I could use some feedback on how to do a few things 

- [ ] How should I deal with a config file that defined ClusterNetworkCIDR the old way?

- [ ] In pkg/sdn/plugin/master.go there where checks to see if the cluster cidr was shrunk, I don't think that check is still required. It should be valid to break a large cluster cidr into it's smaller components. Is checking that all objects are allocated in defined places enough? As done by checkclusterobjects() in pkg/sdn/plugin/common.go

- [x] I need to change oc get clusternetwork to show a comma separated list 

- [ ] finish the unit testing


review of the work so far and input on the above questions are appreciated 
@knobunc @dcbw @danwinship @rajatchopra

danwinship commented:
> addresses can be defined in the master config file by
> 
>     networkConfig:
>       clusterNetworkCIDR: ""
>       clusterNetworkConfig:
>       - clusterNetworkCIDR: 10.128.0.0/28
>       - clusterNetworkCIDR: 12.128.0.0/28

Hm... Is there any reason you did it that way? It seems like it would be nicer to just have

    clusterNetworkCIDRs:
    - 10.128.0.0/28
    - 12.128.0.0/28

aka

    clusterNetworkCIDRs: [ "10.128.0.0/28", "12.128.0.0/28" ]

>     How should I deal with a config file that defined ClusterNetworkCIDR the old way?

Treat it the same as if it was defined the new way, but with only a single CIDR range

>     In pkg/sdn/plugin/master.go there where checks to see if the cluster cidr was shrunk, I don't think that check is still required. It should be valid to break a large cluster cidr into it's smaller components. Is checking that all objects are allocated in defined places enough? As done by checkclusterobjects() in pkg/sdn/plugin/common.go

Yes, that sounds right. Make sure the nodes do the right thing on restart as well.

knobunc commented:
@openshift/networking

knobunc commented:
@danwinship the reason he did it that way is because we will shortly need to accommodate hot-resizing the subnets (or even supporting a way to have different size host subnets per range, perhaps with a way to label the subnets so different nodes can deliberately have different allocations).

pecameron commented on pkg/sdn/plugin/common.go:
```diff
@@ -39,20 +39,36 @@ func clusterNetworkToString(n *osapi.ClusterNetwork) string {
 	return fmt.Sprintf("%s (network: %q, hostSubnetBits: %d, serviceNetwork: %q, pluginName: %q)", n.Name, n.Network, n.HostSubnetLength, n.ServiceNetwork, n.PluginName)
 }
 
+//determine if the first argument contains any of the second
```
Isn't this contains the second or is contained in the second?

pecameron commented on pkg/sdn/plugin/common.go:
```diff
@@ -39,20 +39,36 @@ func clusterNetworkToString(n *osapi.ClusterNetwork) string {
+func parseNetworkInfo(clusterNetwork []osapi.ClusterNetworkEntry, serviceNetwork string) (*NetworkInfo, error) {
+	var cn []*net.IPNet
+
+	for _, entry := range clusterNetwork {
```
I think you need to check each cidr with all other cidrs (except for itself). How many cidrs do we support?

pravisankar replied:
Yes, this is needed. Do we want to restrict number of cidrs that we support? If users specify single IPs or sparse cidrs, this could be huge.

JacobTanenbaum replied:
What do you mean check each cidr with all other cidrs?

pravisankar replied:
This is to verify there is no overlap between the given cidrs, otherwise pods may end up with same IPs.

pecameron commented on pkg/sdn/plugin/common.go:
```diff
@@ -116,8 +134,8 @@ func (ni *NetworkInfo) checkClusterObjects(subnets []osapi.HostSubnet, pods []ka
-		} else if !ni.ClusterNetwork.Contains(subnetIP) {
-			errList = append(errList, fmt.Errorf("existing node subnet: %s is not part of cluster network: %s", subnet.Subnet, ni.ClusterNetwork.String()))
+		} else if _, contains := cidrListContains(ni.ClusterNetwork, subnetIP); !contains {
+			errList = append(errList, fmt.Errorf("existing node subnet: %s in not part of any cluster network CIDR", subnet.Subnet))
```
"in" ?

pecameron commented on pkg/sdn/plugin/common.go:
```diff
@@ -156,7 +174,7 @@ func getNetworkInfo(osClient *osclient.Client) (*NetworkInfo, error) {
 	}
 
-	return parseNetworkInfo(cn.Network, cn.ServiceNetwork)
+	return parseNetworkInfo(cn.ClusterDef, cn.ServiceNetwork)
```
Do we care about externalIP and load balancer IP here?

danwinship commented on pkg/cmd/server/api/types.go:
```diff
@@ -610,10 +610,11 @@ type UserAgentDenyRule struct {
-	ServiceNetworkCIDR string
+	NetworkPluginName    string
+	ClusterNetworkCIDR   string
+	ClusterNetworkConfig []ClusterNetworkEntry
```
I'd probably go with just `ClusterNetworks`.

It would be nice if we could get rid of the old `ClusterNetworkCIDR` field, but I'm not sure how backward-compat works with startup config objects. Old master-config.yaml files need to keep working though, so whatever that requires...

danwinship commented on pkg/cmd/server/api/types.go:
```diff
@@ -626,6 +627,10 @@ type MasterNetworkConfig struct {
 }
 
+type ClusterNetworkEntry struct {
+	ClusterNetworkCIDR string
```
Maybe just `CIDR`? You've already got `ClusterNetwork` in the struct name and field name...

Also, if the plan is to eventually have per-network HostSubnetLength values, I think we should add that here now; if we don't support different values yet, you can just add code to make sure every ClusterNetworkEntry has the same value.

(Also, if we're tweaking HostSubnetLength handling, we might want to consider renaming it and flipping the math around; the fact that it counts bits from the end of the IP address rather than the start is confusing. (Eg, HostSubnetLength of 9 means each node gets a  /23 subnet (`32 - 9 = 23`). I think it would be clearer for most users if the config option was "23" rather than "9".

pravisankar replied:
+1 on CIDR.
If you plan to do global/local HostSubnetLength field that I was suggesting, then.
```
type ClusterNetworkEntry struct {
      ClusterNetworkCIDR string
      HostSubnetLength     uint32
}
```

danwinship commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -635,7 +635,8 @@ func newAdmissionChain(pluginNames []string, admissionConfigFilename string, plu
 			// we need to set some customer parameters, so create by hand
-			restrictedNetworks, err := serviceadmit.ParseSimpleCIDRRules([]string{options.NetworkConfig.ClusterNetworkCIDR, options.NetworkConfig.ServiceNetworkCIDR})
+			// PLACEHOLDER HAVE TO FIGURE OUT OUT TO DO
+			restrictedNetworks, err := serviceadmit.ParseSimpleCIDRRules([]string{options.NetworkConfig.ServiceNetworkCIDR})
```
You need to pass an array containing all of the clusternetwork cidr values and the servicenetwork cidr value

danwinship commented on pkg/sdn/api/v1/types.go:
```diff
@@ -20,6 +20,8 @@ type ClusterNetwork struct {
 	// Network is a CIDR string specifying the global overlay network's L3 space
 	Network string `json:"network" protobuf:"bytes,2,opt,name=network"`
+	// PLACEHOLDER
+	ClusterDef []ClusterNetworkEntry `json:"clusterDef" protobuf:"bytes,6,rep,name=clusterDef"`
```
https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md has some information on changing kubernetes API objects, and in particular discusses the backward-compatibility snafus that come with trying to turn a single-element field into an array... (The "API Conventions" link near the top of that document is also filled with useful information about how kubernetes APIs work.)

The easiest way to make this work backward-compatibly would be to say that the "default"/first network is still specified by the `Network` and `HostSubnetLength` fields, but the 2nd to Nth networks are part of a new array field (which could be called `AdditionalNetworks` or something like that).

You need a doc comment above the field, and between the doc comment and the actual declaration should be

    // +optional

to indicate that it's an optional field

danwinship commented on pkg/sdn/api/v1/types.go:
```diff
@@ -28,6 +30,11 @@ type ClusterNetwork struct {
 
+// PLACEHOLDER
+type ClusterNetworkEntry struct {
+	ClusterNetworkCIDR string `json:"clusterNetworkCIDR" protobuf:"bytes,1,opt,name=clusterNetworkCIDR"`
```
(same comments as above)

danwinship commented on pkg/sdn/api/types.go:
```diff
@@ -16,11 +16,16 @@ type ClusterNetwork struct {
 	metav1.ObjectMeta
 
 	Network          string
+	ClusterDef       []ClusterNetworkEntry
```
The "unversioned" ClusterNetwork type (the one in pkg/sdn/api/types.go, as opposed to pkg/sdn/api/v1/types.go) is an internal-only object, not part of the API, so it is not subject to backward-compatibility rules. So you could get rid of the old Network and HostSubnetLength fields here and only have the new array.

The one catch with this is that if the unversioned type and the v1 type aren't identical, then you'll need to add some hand-written conversion functions, but this isn't difficult. There are no examples in pkg/sdn/api but you can see examples in other API groups (eg, pkg/image/api/v1/conversion.go).

danwinship commented on pkg/sdn/api/types.go:
```diff
@@ -16,11 +16,16 @@ type ClusterNetwork struct {
 }
 
+type ClusterNetworkEntry struct {
+	ClusterNetworkCIDR string
```
same comments as in pkg/cmd/server/api/types.go above

pravisankar commented:
With multiple cluster network CIDR support, hostsubnet length don't need to be same for each CIDR. We may not add this support immediately but it will be nice if we can make the config compatible with the future change that is more likely to be added.

How about this master config? 
```
clusterNetworkCIDR:   (deprecate this one)
hostSubnetLength: 2   (global)
clusterNetworkCIDRs:
    - CIDR: 10.128.0.0/28
      hostSubnetLength: 2   (local)
    - CIDR: 12.128.0.0/28
      hostSubnetLength: 2
```

- If local hostSubnetLength field is not set, apply the global hostSubnetLength (during default conversions)
- If global and any of the local hostSubnetLength fields (if set) are not matching, bail out with an error similar to 'different host subnet lengths for cluster network cidrs are not currently supported'.

How should I deal with a config file that defined ClusterNetworkCIDR the old way?

 I have deprecated/renamed fields in master/node config before (retaining backward compatibility),
 you can check these PRs for reference:
 https://github.com/openshift/origin/commit/6557c81c3b6c34bb3e579d0cbfee72fb2ee0175e
 https://github.com/openshift/origin/pull/4101/commits/4663c0891040eb36ea09daff6f3eff407cd7d7a1

pravisankar commented on pkg/sdn/plugin/sdn_controller.go:
```diff
@@ -160,11 +160,11 @@ func (plugin *OsdnNode) SetupSDN() (bool, error) {
+	//err = ioutil.WriteFile("/run/openshift-sdn/config.env", []byte(config), 0644)
+	//if err != nil {
+	//	return false, err
+	//}
```
This was discussed in https://github.com/openshift/origin/pull/13726

pravisankar commented on pkg/cmd/server/start/master_args.go:
```diff
@@ -188,6 +189,15 @@ func (args MasterArgs) BuildSerializeableMasterConfig() (*configapi.MasterConfig
+		}
+		clusterNetworkConfig = append(clusterNetworkConfig, clusterNetworkEntry)
+	}
+
```
My understanding is command line arguments is for simple/essential use cases and complex use cases need to be specified in configuration file. If we treat multiple cidr use case as complex then we don't need to support command line args. 
Supporting comma separated list here makes it harder to specify different hostsubnetLength for each cidr later.

pravisankar commented on pkg/sdn/api/v1/types.go:
```diff
@@ -20,6 +20,8 @@ type ClusterNetwork struct {
 	// Network is a CIDR string specifying the global overlay network's L3 space
 	Network string `json:"network" protobuf:"bytes,2,opt,name=network"`
+	// PLACEHOLDER
+	ClusterDef []ClusterNetworkEntry `json:"clusterDef"`
```
Needs to add protobuf tag

pravisankar commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -20,38 +20,67 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
 
+//intersect tests two CIDR addresses to see if the first contains the second
+func intersect(cidr1, cidr2 string) bool {
+	_, net1, _ := net.ParseCIDR(cidr1)
```
Check the error (similar to cidr2 below)

pravisankar commented on pkg/sdn/plugin/master.go:
```diff
@@ -178,3 +190,20 @@ func clusterNetworkChanged(obj *osapi.ClusterNetwork, old *osapi.ClusterNetwork)
 }
+
+func clusterDefEquals(obj, old []osapi.ClusterNetworkEntry) bool {
+	if obj == nil && old == nil {
```
Not needed, condition below covers this check.

pravisankar commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -996,8 +996,12 @@ func printNetNamespaceList(list *sdnapi.NetNamespaceList, w io.Writer, opts kpri
+	}
 	name := formatResourceName(opts.Kind, n.Name, opts.WithKind)
-	_, err := fmt.Fprintf(w, "%s\t%s\t%d\t%s\t%s\n", name, n.Network, n.HostSubnetLength, n.ServiceNetwork, n.PluginName)
+	_, err := fmt.Fprintf(w, "%s\t%s\t%d\t%s\t%s\n", name, strings.Join(cidrList, ","), n.HostSubnetLength, n.ServiceNetwork, n.PluginName)
```
Yes, it will hard to read when there are several cidrs. May be one row for each cidr?
```
for _, cidr := range n.ClusterDef {
   , err := fmt.Fprintf(w, "%s\t%s\t%d\t%s\t%s\n", name, cidr. ...)
}
```

JacobTanenbaum commented:
@danwinship @knobunc @pravisankar Should I change the math for hostSubnetLength field? Won't that How could we upgrade a config from the old math to the new math for hostsubnetlength?

pravisankar commented:
From the stand point of user using multiple cidr, with the current implementation user may not use a cidr because it conflicts with global HostSubnetLength (they don't have enough bits in the cidr to allocate for the pod). I feel cidr and hostSubnetLength go hand in hand. Why don't we work on supporting multiple cidr and hostsubnetLength per cidr? This will remove some of these limitations and will avoid unnecessary migration later on.

@openshift/networking

JacobTanenbaum commented:
@danwinship @pravisankar is this moving in the right direction? I deprecated the global hostsubnetlength field and added one per cidr. I also changed it so only one could be defined on the cli and others need to be added to the config

pecameron commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -24,37 +24,50 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+		} else {
+			maskLen, addrLen := clusterIPNet.Mask.Size()
+			if cn.HostSubnetLength > uint32(addrLen-maskLen) {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), cn.HostSubnetLength, "subnet length is too large for clusterNetwork"))
```
The message is not all that helpful if you have a complex setup with lots of cidrs.

pecameron commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -24,37 +24,50 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+		}
+		if (serviceIPNet != nil) && (clusterIPNet != nil) && serviceIPNet.Contains(clusterIPNet.IP) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, "cluster network overlaps with service network"))
+		}
```
Can external IPs and loadbalancer IPs overlap the above?
This doesn't check for overlaps in the set of ClusterNetworks.

pravisankar replied:
clusterNetwork overlap with external CIDR or ingress CIDR is not checked (looks like we missed this issue before).
May be we should fix this one now?

pravisankar commented on pkg/sdn/plugin/common_test.go:
```diff
@@ -138,31 +151,39 @@ func Test_checkClusterObjects(t *testing.T) {
-				ClusterNetwork: []*net.IPNet{mustParseCIDR("1.2.3.0/24")},
+				ClusterNetworks: []ClusterNetwork{
+					{ClusterCIDR: mustParseCIDR("1.2.3.0/24"), HostSubnetLength: 8},
+				},
```
Same here..need to add test cases that has multiple cidr/hostsubnetlength.

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -36,8 +36,9 @@ func (master *OsdnMaster) SubnetStartMaster(clusterNetwork []*net.IPNet, hostSub
-	for _, cidr := range clusterNetwork {
-		subnetAllocator, err := netutils.NewSubnetAllocator(cidr.String(), hostSubnetLength, subrange)
+	for _, cn := range clusterNetworks {
+		log.Infof("KEYWORD: ClusterCIDR: %s, HostSubnetLength: %d", cn.ClusterCIDR.String(), cn.HostSubnetLength)
```
I don't think we want to log here. I guess, you added this log entry for your testing?

JacobTanenbaum commented on pkg/sdn/apis/network/validation/validation_test.go:
```diff
@@ -156,55 +145,50 @@ func TestSetDefaultClusterNetwork(t *testing.T) {
-				ServiceNetwork:   "172.30.0.0/16",
-				PluginName:       "redhat/openshift-ovs-subnet",
+				ObjectMeta:      metav1.ObjectMeta{Name: sdnapi.ClusterNetworkDefault},
+				ClusterNetworks: []sdnapi.ClusterNetworkEntry{{CIDR: "10.20.0.0/16", HostSubnetLength: 8}},
```
This section is testing the ValidateClusterNetwork which validates the default clusternetwork and ensures that you cannot modify the values of the default clusternetwork via the api. I did add a test which adds attempts to add a network which fails and there was one that tries to change the hostsubnet field witch also fails. Do you want me to add another test to ValidateClusterNetwork which tests if the default ClusterNetwork starts with multiple cidr/hostsubnetlengh? There are tests for parsing mutliple cidr/hostsubnet in pkg/sdn/plugin/common_test.go.

pravisankar replied:
Okay, then we don't need multiple cidr test cases here.

JacobTanenbaum commented on pkg/sdn/apis/network/v1/types.go:
```diff
@@ -28,6 +30,12 @@ type ClusterNetwork struct {
 	PluginName string `json:"pluginName,omitempty" protobuf:"bytes,5,opt,name=pluginName"`
 }
 
+// PLACEHOLDER
```
As of now the system fills each subnet from first entry (top) to last, in the future there are plans to label the subnets and label the nodes and allocate based on that label but that does not exist in this PR.

JacobTanenbaum commented on pkg/sdn/apis/network/v1/types.go:
```diff
@@ -20,6 +20,8 @@ type ClusterNetwork struct {
 
 	// Network is a CIDR string specifying the global overlay network's L3 space
 	Network string `json:"network" protobuf:"bytes,2,opt,name=network"`
+	// PLACEHOLDER
```
HostSubnetLength is part of ClusterNEtworkEntry now. 

The size of each HostSubnetLength is compared to its ClusterCIDR in
pkg/sdn/pluginapis/network/validation/validation.go in ValidateClusterNetwork()

JacobTanenbaum commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -495,10 +495,12 @@ type RoutingConfig struct {
-	ClusterNetworkCIDR string `json:"clusterNetworkCIDR"`
-	// HostSubnetLength is the number of bits to allocate to each host's subnet e.g. 8 would mean a /24 network on the host
-	HostSubnetLength uint32 `json:"hostSubnetLength"`
+	// Deprecated and maintained for backwards compatibility, use ClusterNetworks instead
```
I do not believe there is a good way to extend the current implementation.

knobunc commented:
/retest

enj commented:
@JacobTanenbaum took me a while to get the JSON correct ...

```patch
diff --git test/integration/etcd_storage_path_test.go test/integration/etcd_storage_path_test.go
index 5e51df3..b5dce72 100644
--- test/integration/etcd_storage_path_test.go
+++ test/integration/etcd_storage_path_test.go
@@ -282,11 +282,11 @@ var etcdStorageData = map[schema.GroupVersionResource]struct {
 		expectedGVK:      gvkP("", "v1", "HostSubnet"), // expect the legacy group to be persisted
 	},
 	gvr("", "v1", "clusternetworks"): {
-		stub:             `{"metadata": {"name": "cn1"}, "network": "192.168.0.0/24", "hostsubnetlength": 4, "serviceNetwork": "192.168.1.0/24"}`,
+		stub:             `{"metadata": {"name": "cn1"}, "network": "192.168.0.0/24", "hostsubnetlength": 4, "serviceNetwork": "192.168.1.0/24", "clusterNetworks": [{"CIDR": "192.166.0.0/16", "hostsubnetlength": 8}]}`,
 		expectedEtcdPath: "openshift.io/registry/sdnnetworks/cn1",
 	},
 	gvr("network.openshift.io", "v1", "clusternetworks"): {
-		stub:             `{"metadata": {"name": "cn1g"}, "network": "192.168.0.0/24", "hostsubnetlength": 4, "serviceNetwork": "192.168.1.0/24"}`,
+		stub:             `{"metadata": {"name": "cn1g"}, "network": "192.168.0.0/24", "hostsubnetlength": 4, "serviceNetwork": "192.168.1.0/24", "clusterNetworks": [{"CIDR": "192.167.0.0/16", "hostsubnetlength": 8}]}`,
 		expectedEtcdPath: "openshift.io/registry/sdnnetworks/cn1g",
 		expectedGVK:      gvkP("", "v1", "ClusterNetwork"), // expect the legacy group to be persisted
 	},
```

enj commented:
@JacobTanenbaum also, don't you need to make a new version of the type?  The change looks backwards incompatible...

@openshift/api-review

knobunc commented:
/assign @liggitt

JacobTanenbaum commented:
@enj I don't think I need to make a new version of the clusternetworks type I only added to it nothing was removed. @liggitt is that right?

pravisankar commented on pkg/cmd/server/api/serialization_test.go:
```diff
@@ -112,13 +112,28 @@ func fuzzInternalObject(t *testing.T, forVersion schema.GroupVersion, item runti
-				if !(configapi.CIDRsOverlap(cidr, obj.NetworkConfig.ClusterNetworkCIDR) || configapi.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR)) {
+				setCIDR := true
+				for _, clusterNetwork := range obj.NetworkConfig.ClusterNetworks {
+					if configapi.CIDRsOverlap(cidr, clusterNetwork.CIDR) || configapi.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR) {
```
This check `configapi.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR)` can be outside of the `for` loop, it is not dependent on clusterNetwork.

pravisankar replied:
Don't we want to check overlap with CIDRs reserved for external IPs?

pravisankar commented on pkg/cmd/server/api/v1/conversions.go:
```diff
@@ -90,14 +90,23 @@ func SetDefaults_MasterConfig(obj *MasterConfig) {
-		if !(internal.CIDRsOverlap(cidr, obj.NetworkConfig.ClusterNetworkCIDR) || internal.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR)) {
+		cidrOverlap := false
+		for _, entry := range obj.NetworkConfig.ClusterNetworks {
+			if internal.CIDRsOverlap(cidr, entry.CIDR) || internal.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR) {
```
Same as above, `internal.CIDRsOverlap(cidr, obj.NetworkConfig.ServiceNetworkCIDR)` can be checked outside the `for` loop.

pravisankar commented on pkg/cmd/server/api/v1/conversions.go:
```diff
@@ -90,14 +90,23 @@ func SetDefaults_MasterConfig(obj *MasterConfig) {
 	}
+	if len(obj.NetworkConfig.ClusterNetworks) == 0 {
+		obj.NetworkConfig.ClusterNetworks = []ClusterNetworkEntry{{CIDR: obj.NetworkConfig.DeprecatedClusterNetworkCIDR, HostSubnetLength: obj.NetworkConfig.DeprecatedHostSubnetLength}}
+	}
```
When both `ClusterNetworks` and `DeprecatedClusterNetworkCIDR/DeprecatedHostSubnetLength` are defined, it would be nice if can we log a warning. Something like 'Ignoring clusterNetworkCIDR/hostSubnetLength fields in favor of clusterNetwork'.

pravisankar commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -25,36 +25,43 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+		} else {
+			maskLen, addrLen := clusterIPNet.Mask.Size()
+			if cn.HostSubnetLength > uint32(addrLen-maskLen) {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), cn.HostSubnetLength, "subnet length is too large for clusterNetwork"))
```
Since we can have multiple cluster network cidrs, we should print clusterNetwork cidr associated with the hostSubnetLength in case of validation error to make it un-ambiguous.

pravisankar commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -25,36 +25,43 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, err.Error()))
+		}
+		if (clusterIPNet != nil) && (serviceIPNet != nil) && clusterIPNet.Contains(serviceIPNet.IP) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, "service network overlaps with cluster network"))
```
Same here, print the clusterNetwork cidr.

pravisankar commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -25,36 +25,43 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, "service network overlaps with cluster network"))
+		}
+		if (serviceIPNet != nil) && (clusterIPNet != nil) && serviceIPNet.Contains(clusterIPNet.IP) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, "cluster network overlaps with service network"))
```
Same here, print the clusterNetwork cidr.

pravisankar commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -25,36 +25,43 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), clusterNet.HostSubnetLength, "cannot change the default ClusterNetwork record via API."))
+		if len(clusterNet.ClusterNetworks) == len(defaultClusterNetwork.ClusterNetworks) {
+			for index := range defaultClusterNetwork.ClusterNetworks {
+				if clusterNet.ClusterNetworks[index].CIDR != defaultClusterNetwork.ClusterNetworks[index].CIDR || clusterNet.ClusterNetworks[index].HostSubnetLength != defaultClusterNetwork.ClusterNetworks[index].HostSubnetLength {
```
We do not want to bail out when the order of cluster networks changed. 
Lets say, OldClusterNetwork=[cidrA, cidrB] and newClusterNetwork=[cidrB, cidrA]. Here cidrA and cidrB order has changed but the config itself did not change, so there should not be any error in this case.

pravisankar commented on pkg/sdn/apis/network/validation/validation.go:
```diff
@@ -25,36 +25,43 @@ func SetDefaultClusterNetwork(cn sdnapi.ClusterNetwork) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("ClusterNetworks"), clusterNet.ClusterNetworks, "cannot change the default ClusterNetwork record via API"))
 		}
 		if clusterNet.ServiceNetwork != defaultClusterNetwork.ServiceNetwork {
 			allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, "cannot change the default ClusterNetwork record via API."))
```
All these validations are between clusterNetwork and other resources like serviceNetwork, etc. but we should also check overlap among the multiple clusterNetwork cidrs.

JacobTanenbaum replied:
@pravisankar This condition is checked elsewhere... in parseNetworkInfo() (pkg/sdn/plugin/common.go). I thought it was more appropriate there as the clusterNetwork stucts are being created. Should it be moved?

pravisankar commented on pkg/sdn/apis/network/validation/validation_test.go:
```diff
@@ -19,90 +19,98 @@ func TestValidateClusterNetwork(t *testing.T) {
-				ServiceNetwork:   "10.0.0.0/8",
+				ObjectMeta:      metav1.ObjectMeta{Name: "any"},
+				ClusterNetworks: []sdnapi.ClusterNetworkEntry{{CIDR: "10.20.0.0/16", HostSubnetLength: 8}},
+				ServiceNetwork:  "10.0.0.0/8",
```
Add another test case where there is overlap among cidrs in cluster networks.

pravisankar commented on pkg/sdn/plugin/common.go:
```diff
@@ -31,20 +31,60 @@ func clusterNetworkToString(n *osapi.ClusterNetwork) string {
+		}
+		cns = append(cns, ClusterNetwork{ClusterCIDR: clusterAddress, HostSubnetLength: entry.HostSubnetLength})
 	}
+
```
You are already checking for overlaps among clusternetowrk cidrs here, then this should be used in ValidateClusterNetwork()?

pravisankar commented on pkg/sdn/plugin/common.go:
```diff
@@ -148,7 +196,7 @@ func getNetworkInfo(osClient *osclient.Client) (*NetworkInfo, error) {
 	}
 
-	return parseNetworkInfo(cn.Network, cn.ServiceNetwork)
+	return parseNetworkInfo(cn.ClusterNetworks, cn.ServiceNetwork)
```
I don't think, we need to check overlap among cluster network cidrs here. 
As commented before, I believe this could be done in ValidateClusterNetwork().

danwinship commented on pkg/sdn/plugin/node_iptables.go:
```diff
@@ -165,23 +152,53 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
+		if n.masqueradeServices {
+			masqRule = []string{"-s", cidr, "-m", "comment", "--comment", "masquerade pod-to-service and pod-to-external traffic", "-j", "MASQUERADE"}
+		} else {
+			masqRule = []string{"-s", cidr, "!", "-d", cidr, "-m", "comment", "--comment", "masquerade pod-to-external traffic", "-j", "MASQUERADE"}
```
This won't work right. The rule is supposed to be "if a packet starts in the SDN and doesn't go back into the SDN, it needs to be masqueraded". But eg, if you had two networks 1.2.3.0/8 and 4.5.6.0/8 then this would generate:

    -A OPENSHIFT-MASQUERADE -s 1.2.3.0/8 ! -d 1.2.3.0/8 -j MASQUERADE
    -A OPENSHIFT-MASQUERADE -s 4.5.6.0/8 ! -d 4.5.6.0/8 -j MASQUERADE

and so if a packet went from 1.2.3.4 to 4.5.6.7, it would end up getting masqueraded (which you didn't want) because it matches the first rule.

This isn't totally trivial to fix. One way would be to have a second table; in OPENSHIFT-MASQUERADE you'd have a series of rules to match packets whose source was in the SDN:

    -A OPENSHIFT-MASQUERADE -s 1.2.3.0/8 -j OPENSHIFT-MASQUERADE-2
    -A OPENSHIFT-MASQUERADE -s 4.5.6.0/8 -j OPENSHIFT-MASQUERADE-2

and then in OPENSHIFT-MASQUERADE-2 you'd bail out if the destination was also in the SDN, but masquerade otherwise:

    -A OPENSHIFT-MASQUERADE-2 -d 1.2.3.0/8 -j RETURN
    -A OPENSHIFT-MASQUERADE-2 -d 4.5.6.0/8 -j RETURN
    -A OPENSHIFT-MASQUERADE-2 -j MASQUERADE

It's totally possible there's a simpler way to do this; I'm not super-duper iptables-savvy. @dcbw might have ideas?

dcbw replied:
@danwinship @JacobTanenbaum I'm probably missing something here, but shouldn't packets that are CIDR->CIDR stay in the SDN and never get to tun0 in the first place?  If a packet gets to tun0 that is meant directly for another pod in the SDN, then our OVS rules have failed and that packet should probably get dropped.

eg, in the above example, I can't see why we'd ever get a packet from 1.2.3.4 -> 4.5.6.7 hitting tun0 and thus iptables.  That should just get handled via OVS rules and send directly to the other node in the VXLAN, not outside the SDN.

Or am I missing something in architecture here that I should have seen...?

danwinship replied:
> I'm probably missing something here, but shouldn't packets that are CIDR->CIDR stay in the SDN and never get to tun0 in the first place?

Yes, you're missing something (though it's totally non-obvious so we probably should have a comment here). This rule applies to packets that go pod -> service IP -> pod; after the iptables service DNAT rule applies, it will have source and destination both in clusterNetworkCIDR(s), but it needs to be masqueraded to ensure that it goes back through iptables again on the way back (to get un-DNATed)

pravisankar commented on pkg/sdn/plugin/sdn_controller.go:
```diff
@@ -87,9 +87,11 @@ func (plugin *OsdnNode) alreadySetUp(localSubnetGatewayCIDR, clusterNetworkCIDR
+		for _, clusterCIDR := range clusterNetworkCIDR {
+			if strings.Contains(route, clusterCIDR+" ") {
+				found = true
+				break
```
We also want to break from routes for loop not just the clusterNetworkCIDR loop?

danwinship replied:
This ends up only checking that at least one clusterNetwork is configured. It needs to check that all of them are.

danwinship commented on pkg/sdn/plugin/node_iptables.go:
```diff
@@ -165,23 +152,53 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
+				srcChain: "POSTROUTING",
+				srcRule:  []string{"-m", "comment", "--comment", "rules for masquerading OpenShift traffic"},
+				rules: [][]string{
+					{"-s", cidr, "-m", "comment", "--comment", "masquerade pod-to-service and pod-to-external traffic", "-j", "MASQUERADE"},
```
this seems like a duplicate / cut+paste error?

danwinship commented on pkg/sdn/plugin/node_iptables.go:
```diff
@@ -165,23 +152,53 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
+		chainArray = append(chainArray,
+			Chain{
+				table:    "nat",
+				name:     "OPENSHIFT-MASQUERADE",
```
You don't want per-network `Chain` objects, you want a single `"OPENSHIFT-MASQUERADE"` chain, with per-network `rules`

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -75,26 +79,29 @@ func (master *OsdnMaster) addNode(nodeName string, nodeIP string, hsAnnotations
+	for _, possibleSubnet := range master.subnetAllocatorList {
+		sn, err := possibleSubnet.GetNetwork()
+		if err == netutils.ErrSubnetAllocatorFull {
+			log.Infof("error allocating network for node %s:%v in subnet %s", nodeName, err)
```
It is not an error when one of the subnet is fully consumed, we just need to assign from rest of the available subnets.
May be:
``` 
if err != nil {
    if err != netutils.ErrSubnetAllocatorFull {
        log.Errorf("Error fetching network ...")
    }
}
// Got valid sn, process further
```

danwinship replied:
I don't think there's any reason to log this unless *all* of the subnets are full. Eg, someone might fill up their original clusterNetwork and then add a second; there's no reason to log a message about the fact that the first network is full every time they add another node, when there's still plenty of room in the second network

danwinship commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -75,26 +79,29 @@ func (master *OsdnMaster) addNode(nodeName string, nodeIP string, hsAnnotations
+			sub, err = master.osClient.HostSubnets().Create(sub)
+			if err != nil {
+				possibleSubnet.ReleaseNetwork(sn)
+				log.Infof("error creating subnet %s for node %s: %v", sn.String(), nodeName, err)
```
This seems like it should probably still be fatal; there's not really any reason why trying to create a HostSubnet on one subnet would fail but the next would succeed...

knobunc commented on pkg/sdn/plugin/master.go:
```diff
@@ -164,27 +166,23 @@ func (master *OsdnMaster) checkClusterNetworkAgainstClusterObjects() error {
-			// oldSize and newSize are, eg the "16" in "10.1.0.0/16", so
-			// "newSize < oldSize" means the new network is larger
-			if !(newSize < oldSize && newNet.Contains(oldBase)) {
-				return true, fmt.Errorf("cannot change clusterNetworkCIDR to a value that does not include the existing network.")
```
I don't think we want the old restriction that the old network must be included in the new.  We definitely want to make sure that the old hostsubnet allocations all fall within the new network (and fail hard if they do not).

JacobTanenbaum commented on pkg/sdn/plugin/common.go:
```diff
@@ -31,20 +31,60 @@ func clusterNetworkToString(n *osapi.ClusterNetwork) string {
-		_, cn, err := net.ParseCIDR(clusterNetwork)
+type ClusterNetwork struct {
+	ClusterCIDR      *net.IPNet
+	HostSubnetLength uint32
```
After my additions the NetworkInfo needs the hostsubnetlength. previously you could pass the global hostsubnetlength to NewSubnetAllocator() but now the hostsubnetlength has to be associated with it's cidr range

JacobTanenbaum commented:
@danwinship @pravisankar please review.

JacobTanenbaum commented:
/test cmd

JacobTanenbaum commented:
/test extended_networking_minimal

JacobTanenbaum commented:
/test check

knobunc commented:
/retest

knobunc commented:
/lgtm

pravisankar commented on pkg/network/master/master.go:
```diff
@@ -166,33 +172,27 @@ func (master *OsdnMaster) checkClusterNetworkAgainstClusterObjects() error {
+
 	if old.PluginName != obj.PluginName {
 		changed = true
 	}
```
We can quickly return in some cases without requiring comparing old and new cluster networks.

```
if old.ServiceNetwork != obj.ServiceNetwork {
  return true, fmt.Errorf(...)
} else if old.PluginName != obj.PluginName {
  return true, nil
} else if len(old.ClusterNetworks) != len(obj.ClusterNetworks) {
  return true, nil
} else {
  // Compare old and new cluster networks and set changed
  return changed, nil
}
return false, nil
```

pravisankar commented on pkg/network/master/subnets.go:
```diff
@@ -76,26 +80,29 @@ func (master *OsdnMaster) addNode(nodeName string, nodeIP string, hsAnnotations
+		sn, err := possibleSubnet.GetNetwork()
+		if err != nil && err != netutils.ErrSubnetAllocatorFull {
+			return "", fmt.Errorf("error allocating network for node %s: %v", nodeName, err)
+		} else {
```
When err != nil and err == netutils.ErrSubnetAllocatorFull, then we will use invalid `sn` here. 
We should do something like:
```
sn, err := possibleSubnet.GetNetwork()
if err ==  netutils.ErrSubnetAllocatorFull {
    // Current subnet exhausted, check next one
   continue
} else if err != nil {
  glog.Errorf("Error allocating network from subnet: %v", posssibleSubnet)
  continue
} else {
  // Process valid subnet
}
```

pravisankar commented on pkg/network/node/sdn_controller.go:
```diff
@@ -89,13 +89,17 @@ func (plugin *OsdnNode) alreadySetUp(localSubnetGatewayCIDR, clusterNetworkCIDR
+			}
+		}
+		if !found {
+			return false
```
```
for _, route := range routes {
  found = false
  for _, clusterCIDR := range clusterNetworkCIDR {
     if route.Dst != nil && route.Dst.String() == clusterCIDR {
           found = true
           break
     }
  }
  if !found {
      return false
  }
}
```

pravisankar commented on pkg/oc/cli/describe/printer.go:
```diff
@@ -1040,7 +1040,24 @@ func printNetNamespaceList(list *networkapi.NetNamespaceList, w io.Writer, opts
+	var networksList []string
+	var networks string
+	for _, cidr := range n.ClusterNetworks {
+		networksList = append(networksList, strings.Join([]string{cidr.CIDR, fmt.Sprintf("%d", cidr.HostSubnetLength)}, ":"))
```
simplify fmt.Sprintf("%s:%d", cidr.CIDR, cidr.HostSubnetLength)

pravisankar commented on pkg/oc/cli/describe/printer.go:
```diff
@@ -1040,7 +1040,24 @@ func printNetNamespaceList(list *networkapi.NetNamespaceList, w io.Writer, opts
+		networks = fmt.Sprintf("%s + %d more...",
+			strings.Join(networksList[:numOfNetworksShown], ", "),
+			len(networksList)-numOfNetworksShown)
+	} else if len(networksList) > 0 {
```
nit: `} else {`

JacobTanenbaum commented on pkg/network/master/master.go:
```diff
@@ -52,7 +51,11 @@ func Start(networkConfig osconfigapi.MasterNetworkConfig, osClient *osclient.Cli
+	for _, cidr := range networkConfig.ClusterNetworks {
+		clusterNetworkEntries = append(clusterNetworkEntries, networkapi.ClusterNetworkEntry{CIDR: cidr.CIDR, HostSubnetLength: cidr.HostSubnetLength})
+	}
+	master.networkInfo, err = common.ParseNetworkInfo(clusterNetworkEntries, networkConfig.ServiceNetworkCIDR)
```
I use clusterNetworkEntries to convert between the ClusterNetworkEntry types. The one defined in the master (pkg/cmd/server/api/types.go) and the internal type for the sdn (pkg/network/apis/network/types.go)

pravisankar replied:
ClusterNetworkEntry struct in pkg/cmd/server/api/types.go and pkg/network/apis/network/types.go are exactly the same thing. Why not define once in pkg/cmd/server/api/types.go and reuse the same struct in pkg/network/apis/network/types.go ? Then you can directly pass networkConfig.ClusterNetworks to ParseNetworkInfo().

JacobTanenbaum replied:
@pravisankar I just tried to make the switch and reuse the code from pkg/cmd/server/api/types.go, it is illegal to import pkg/cmd/server/api/types.go into pkg/network/apis/network/types.go hack/verify-imports.sh complains

pravisankar replied:
@JacobTanenbaum 
I was surprised that we can't reuse the structs in types.go, so I looked at other types.go files.
Gist is that we can import/reuse type structs from other packages as long as there are no import cycles (A->B, B->A or A->B, B->C, C->A). Looks like you can reuse ClusterNetworkEntry network/types.go in api/types.go (just like resource.Quantity, runtime.Object, etc. in api/types.go).
Ideally we should duplicate ClusterNetworkEntry only when they really diverge in the future (and this may not happen). 
This is not critical and don't want to block on this, we can get this pr merged and you can submit a follow up pr.

JacobTanenbaum replied:
@pravisankar I will look at this tonight, if I can't get it in before this PR gets on the merge queue I will try to have this ready right after it lands

JacobTanenbaum replied:
@pravisankar so far only kuberenetes packages are imported into server/api/types.go, I don't know if it on purpose or coincidental so I think that discussion might be better left for a follow up PR

JacobTanenbaum commented:
/retest

pravisankar commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -37,27 +38,24 @@ func ValidateClusterNetwork(clusterNet *networkapi.ClusterNetwork) field.ErrorLi
 		clusterIPNet, err := netutils.ParseCIDRMask(cn.CIDR)
 		if err != nil {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("network"), cn.CIDR, err.Error()))
+			allErrs = append(allErrs, field.Invalid(field.NewPath("clusterNetwprls").Index(i).Child("cidr"), cn.CIDR, err.Error()))
```
typo field.NewPath("clusterNetworks")

danwinship commented on pkg/network/master/master.go:
```diff
@@ -63,10 +66,9 @@ func Start(networkConfig osconfigapi.MasterNetworkConfig, networkClient networkc
 		ObjectMeta: metav1.ObjectMeta{Name: networkapi.ClusterNetworkDefault},
 
-		Network:          networkConfig.ClusterNetworkCIDR,
-		HostSubnetLength: networkConfig.HostSubnetLength,
```
for API compatibility, we still need to set ClusterNetworkCIDR and HostSubnetLength, at least in the case where there is only a single network, so that code that relies on being able to read those values from the default ClusterNetwork will still work.

danwinship replied:
This comment still applies

danwinship commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -24,37 +26,42 @@ func SetDefaultClusterNetwork(cn networkapi.ClusterNetwork) {
 	allErrs := validation.ValidateObjectMeta(&clusterNet.ObjectMeta, false, path.ValidatePathSegmentName, field.NewPath("metadata"))
+	var testedCIDRS []*net.IPNet
 
-	clusterIPNet, err := netutils.ParseCIDRMask(clusterNet.Network)
```
Should validate that either Network and HostSubnetLength are unset, or that they are identical to the first element of ClusterNetworks.

JacobTanenbaum replied:
That happens in the master config validation/conversion. If both ClusterNetworks/Hostsubnet are set and ClusterNetworks are set Clayton wants startup to fail and report that as an error

danwinship replied:
That's the config object. This is the API object. They're separate.

Although what I said here will do the wrong thing for user-created non-"default" ClusterNetwork objects. So... ok, we should validate that if Network and/or HostSubnetLength is set, then they are valid values, but we don't need to compare them against ClusterNetworks.

Oh, except that if ClusterNetworks is 0-length, then Network/HostSubnetLength need to be set.

JacobTanenbaum replied:
I never set the api object for network or hostsubnetlengh, in pkg/cmd/server/api/v1/conversions.go I use DeprecatedHostSubnetLength and DeprecatedClusterNetwork to build a clusterNetworkEntry if len(clusterNetworkEntry) = 0 and unset the deprecated ones.

danwinship commented on pkg/network/common/common.go:
```diff
@@ -110,8 +130,8 @@ func (ni *NetworkInfo) CheckClusterObjects(subnets []networkapi.HostSubnet, pods
-		} else if !ni.ClusterNetwork.Contains(subnetIP) {
-			errList = append(errList, fmt.Errorf("existing node subnet: %s is not part of cluster network: %s", subnet.Subnet, ni.ClusterNetwork.String()))
+		} else if _, contains := ClusterNetworkListContains(ni.ClusterNetworks, subnetIP); !contains {
+			errList = append(errList, fmt.Errorf("existing node subnet: %s in not part of any cluster network CIDR", subnet.Subnet))
```
typo "**in** not part of"

danwinship commented on pkg/network/master/master.go:
```diff
@@ -102,6 +104,10 @@ func Start(networkConfig osconfigapi.MasterNetworkConfig, networkClient networkc
 				configCN.TypeMeta = existingCN.TypeMeta
 				configCN.ObjectMeta = existingCN.ObjectMeta
+				if err = master.checkClusterNetworkAgainstClusterObjects(); err != nil {
+					log.Errorf("ERROR: Attempting to create cluster that excludes existing objects: %v", err)
```
"Attempting to modify cluster to exclude existing objects"

danwinship commented on pkg/network/node/iptables.go:
```diff
@@ -167,23 +153,56 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
+				table:    "nat",
+				name:     "OPENSHIFT-MASQUERADE-2",
+				srcChain: "POSTROUTING",
+				srcRule:  []string{"-m", "comment", "--comment", "additional rules for masquerading Openshift traffic"},
```
This will create a rule in POSTROUTING jumping to OPENSHIFT-MASQUERADE-2. You want to leave srcChain and srcRule unset here (but you'll have to also update the code in `syncIPTablesRules` to handle that case, where it has to make sure the chain exists, but not create a rule jumping to it).

danwinship commented on pkg/oc/cli/describe/printer.go:
```diff
@@ -62,7 +62,7 @@ var (
 	hostSubnetColumns          = []string{"NAME", "HOST", "HOST IP", "SUBNET"}
 	netNamespaceColumns        = []string{"NAME", "NETID"}
-	clusterNetworkColumns      = []string{"NAME", "NETWORK", "HOST SUBNET LENGTH", "SERVICE NETWORK", "PLUGIN NAME"}
+	clusterNetworkColumns      = []string{"NAME", "NETWORK:HOST SUBNET LENGTH", "SERVICE NETWORK", "PLUGIN NAME"}
```
just call it "CLUSTER NETWORKS"

pravisankar commented:
@JacobTanenbaum needs gofmt on ./pkg/network/apis/network/types.go

pravisankar commented:
Swagger spec failure:
```
[FATAL] Swagger spec is out of date. Please run hack/update-generated-swagger-spec.sh:
[FATAL] diff -Naupr -I 'Auto generated by' /go/src/github.com/openshift/origin/api/protobuf-spec/github_com_openshift_origin_pkg_network_apis_network_v1.proto /go/src/github.com/openshift/origin/_output/verify-generated-swagger-spec/api/protobuf-spec/github_com_openshift_origin_pkg_network_apis_network_v1.proto
[FATAL] --- /go/src/github.com/openshift/origin/api/protobuf-spec/github_com_openshift_origin_pkg_network_apis_network_v1.proto	2017-09-25 21:07:08.000000000 +0000
[FATAL] +++ /go/src/github.com/openshift/origin/_output/verify-generated-swagger-spec/api/protobuf-spec/github_com_openshift_origin_pkg_network_apis_network_v1.proto	2017-09-25 21:20:19.003756991 +0000
[FATAL] @@ -39,7 +39,7 @@ message ClusterNetwork {
[FATAL]  // Contains the CIDR and the host subnet length that defines this particular ClusterNetworkEntry
[FATAL]  message ClusterNetworkEntry {
[FATAL]    // CIDR is the CIDR string specifying this particular ClusterNetworkEntry
[FATAL] -  optional string CIDR = 1;
[FATAL] +  optional string cidr = 1;
[FATAL]  
[FATAL]    // HostSubnetLength is the number of bits of the accompanying CIDR address to allocate to each node. eg, 8 would mean that each node would have a /24 slice of the overlay network for its pods
[FATAL]    optional uint32 hostsubnetlength = 2;
[ERROR] hack/verify-generated-swagger-spec.sh exited with code 1 after 00h 00m 33s
make: *** [verify] Error 1
[ERROR] PID 13906: hack/lib/build/environment.sh:161: `return "${exitcode}"` exited with status 2.
```

pravisankar commented:
Also need to fix this test case:
```
--- FAIL: TestMasterConfig (0.06s)
	types_test.go:715: Diff:
...
```
https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_check/4867/consoleFull#-209085362356a6671ee4b089752bb4314a

JacobTanenbaum commented:
/test check

JacobTanenbaum commented:
/test cmd

JacobTanenbaum commented:
/test cmd

danwinship commented on pkg/network/apis/network/v1/types.go:
```diff
@@ -21,17 +21,27 @@ type ClusterNetwork struct {
+	// CIDR is the CIDR string specifying this particular ClusterNetworkEntry
+	CIDR string `json:"CIDR" protobuf:"bytes,1,opt,name=cidr"`
+	// HostSubnetLength is the number of bits of the accompanying CIDR address to allocate to each node. eg, 8 would mean that each node would have a /24 slice of the overlay network for its pods
+	HostSubnetLength uint32 `json:"hostSubnetLength" protobuf:"varint,2,opt,name=hostsubnetlength"`
```
you fixed the case of the json name but not the protobuf name

danwinship commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -24,37 +26,58 @@ func SetDefaultClusterNetwork(cn networkapi.ClusterNetwork) {
 		}
 	}
 
+	if len(clusterNet.ClusterNetworks) == 0 {
```
and Network isn't set

knobunc commented:
@smarterclayton @liggitt This is the PR that we spoke about the API change on IRC.  To recap: Jake extended our ClusterNetwork object to add in an array named ClusterNetworks of objects containing CIDRs and SubnetLength entries.  The original fields are left in place in the object and are used if no ClusterNetworks entry is present.

If both the original fields and the new are defined in the config, then the master won't start since the config is invalid.

Can you please give this an API review?  Thanks!

danwinship commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -24,37 +26,58 @@ func SetDefaultClusterNetwork(cn networkapi.ClusterNetwork) {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, "cannot change the default ClusterNetwork record via API."))
-		}
-		if clusterNet.HostSubnetLength != defaultClusterNetwork.HostSubnetLength {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), clusterNet.HostSubnetLength, "cannot change the default ClusterNetwork record via API."))
```
Leave these checks in... if the fields are unset the checks are no-ops anyway, and if they're not unset, then the checks are necessary

pravisankar commented on pkg/cmd/server/api/validation/master.go:
```diff
@@ -842,3 +845,15 @@ func ValidateIngressIPNetworkCIDR(config *api.MasterConfig, fldPath *field.Path)
+func ValidateDeprecatedClusterNetworkConfig(config *api.MasterConfig, fldPath *field.Path) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if len(config.NetworkConfig.ClusterNetworks) > 0 && config.NetworkConfig.DeprecatedHostSubnetLength != 0 {
```
I'm hoping these validations are executed before we call conversions on ClusterNetworks

smarterclayton commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -495,10 +495,12 @@ type RoutingConfig struct {
-	HostSubnetLength uint32 `json:"hostSubnetLength"`
+	// Deprecated and maintained for backwards compatibility, use ClusterNetworks instead
+	DeprecatedClusterNetworkCIDR string `json:"clusterNetworkCIDR,omitempty`
+	// ClusterNetworks is a list of cluster networks that defines the global overlay network's L3 space
```
You need to update the documentation to describe the format.  Godoc on fields IS our API doc.

The field names are fine, but you need to update the godoc to be correct and complete before you get approved.

pravisankar commented on pkg/network/apis/network/types.go:
```diff
@@ -17,12 +17,18 @@ type ClusterNetwork struct {
 
+	ClusterNetworks  []ClusterNetworkEntry
 	Network          string
 	HostSubnetLength uint32
```
We need to keep old Network and HostSubnetLength fields for backward compatibility? (support config written by 3.7 openshift master and read by 3.6 openshift master) If yes, as part of conversions did we populate these values with ClusterNetworks[0]?

JacobTanenbaum replied:
No, these values are unpopulated. They are only used in the sdn and I removed all use of Network and HostSubnetLength (except for validating if manually created that the values are valid)

pravisankar replied:
Just got confirmation from Clayton that we support config written by 3.7 master to be read by 3.6 master. To allow this, we also need to add ClusterNetworks[0] to DeprecatedNetwork/DeprecatedHostSubnetLength conversion.

pravisankar replied:
@JacobTanenbaum Ignore my previous comment, we are not changing user config and we should not change user config automatically with ansible. Field to List conversion is good and we don't need the reverse conversion.

liggitt commented:
> add in an array named ClusterNetworks of objects containing CIDRs and SubnetLength entries. The original fields are left in place in the object and are used if no ClusterNetworks entry is present.
> 
> If both the original fields and the new are defined in the config, then the master won't start since the config is invalid.

this is typically where our external/internal division helps us. internally, we now deal with a list of cluster networks... internally, you can just deal with a list, and in conversion, you can move the single field (if non-empty) into the list

JacobTanenbaum commented:
@liggitt that is the way it works, if the cluster network is defined the old way then it is converted into the list and used internally as a list of one item. @smarterclayton wanted start up to fail if both methods where used

pravisankar commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -56,6 +82,9 @@ func ValidateClusterNetwork(clusterNet *networkapi.ClusterNetwork) field.ErrorLi
 		if clusterNet.HostSubnetLength != defaultClusterNetwork.HostSubnetLength {
 			allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), clusterNet.HostSubnetLength, "cannot change the default ClusterNetwork record via API."))
 		}
+		if !reflect.DeepEqual(clusterNet.ClusterNetworks, defaultClusterNetwork.ClusterNetworks) {
```
cluster networks could be same but the order in the list could be different?

JacobTanenbaum replied:
A change in the order of the cluster network at this point changes which chunk subnet allocators are created, so I thought it would be better to have it consistant and only change when master is restarted

pravisankar commented on pkg/network/common/common.go:
```diff
@@ -74,8 +95,8 @@ func (ni *NetworkInfo) ValidateNodeIP(nodeIP string) error {
 
-	if ni.ClusterNetwork.Contains(ipaddr) {
-		return fmt.Errorf("node IP %s conflicts with cluster network %s", nodeIP, ni.ClusterNetwork.String())
+	if conflictingCIDR, found := ClusterNetworkListContains(ni.ClusterNetworks, ipaddr); found == true {
```
nit: `; found {`

pravisankar commented on pkg/network/common/common.go:
```diff
@@ -85,11 +117,14 @@ func (ni *NetworkInfo) ValidateNodeIP(nodeIP string) error {
+			if clusterNetwork.ClusterCIDR.Contains(ipNet.IP) {
+
+				errList = append(errList, fmt.Errorf("host network with IP: %s conflicts with cluster network: %s", ipNet.IP.String(), clusterNetwork.ClusterCIDR.String()))
+			}
```
Can't we use CIDRsOverlap() as suggested earlier?

pravisankar commented on pkg/network/master/master.go:
```diff
@@ -166,38 +172,28 @@ func (master *OsdnMaster) checkClusterNetworkAgainstClusterObjects() error {
+			found := false
+			for _, newCIDR := range obj.ClusterNetworks {
+				if newCIDR.CIDR == oldCIDR.CIDR && newCIDR.HostSubnetLength == oldCIDR.HostSubnetLength {
+					found = true
```
Add `break`

pravisankar commented on pkg/network/master/master.go:
```diff
@@ -166,38 +172,28 @@ func (master *OsdnMaster) checkClusterNetworkAgainstClusterObjects() error {
+					found = true
+				}
+			}
+			if found == false {
```
nit: `!found`

pravisankar commented on pkg/oc/cli/describe/describer.go:
```diff
@@ -1689,10 +1689,15 @@ func (d *ClusterNetworkDescriber) Describe(namespace, name string, settings kpri
+		fmt.Fprintf(out, "----\t------------------\n")
+		msg := "%s\t%d\n"
+		for _, clusterNetwork := range cn.ClusterNetworks {
+			fmt.Fprintf(out, msg, clusterNetwork.CIDR, clusterNetwork.HostSubnetLength)
```
nit: `fmt.Fprintf(out, "%s\t%d\n", clusterNetwork.CIDR, clusterNetwork.HostSubnetLength)` is more readable

pravisankar commented on pkg/oc/cli/describe/printer.go:
```diff
@@ -62,7 +62,7 @@ var (
 	hostSubnetColumns          = []string{"NAME", "HOST", "HOST IP", "SUBNET"}
 	netNamespaceColumns        = []string{"NAME", "NETID"}
-	clusterNetworkColumns      = []string{"NAME", "NETWORK", "HOST SUBNET LENGTH", "SERVICE NETWORK", "PLUGIN NAME"}
+	clusterNetworkColumns      = []string{"NAME", "CLUSTER NETWORKS", "SERVICE NETWORK", "PLUGIN NAME"}
```
Now the output of `oc get clusternetworks` is different, do we need to worry about backward compatibility issues here?
@openshift/api-review ^^

knobunc replied:
@smarterclayton says that you can change `oc get` tables and columns.

smarterclayton commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -543,10 +543,12 @@ type RoutingConfig struct {
+	// Deprecated and maintained for backwards compatibility, use ClusterNetworks instead.
+	DeprecatedClusterNetworkCIDR string `json:"clusterNetworkCIDR,omitempty"`
+	// ClusterNetworks is a list of cluster networks that defines the global overlay network's L3 space
+	ClusterNetworks []ClusterNetworkEntry `json:"clusterNetworks"`
```
You need to update the documentation to describe the format. Godoc on fields IS our API doc.

The field names are fine, but you need to update the godoc to be correct and complete before you get approved.

knobunc replied:
@smarterclayton how is:

DeprecatedClusterNetworkCIDR is the CIDR string to specify the global overlay network's L3 space.  Deprecated, but maintained for backwards compatibility, use ClusterNetworks instead.

DeprecatedHostSubnetLength is the number of bits to allocate to each host's subnet e.g. 8 would mean a /24 network on the host.  Deprecated, but maintained for backwards compatibility, use ClusterNetworks instead.

ClusterNetworks is a list of ClusterNetwork objects that defines the global overlay network's L3 space by specifying a set of CIDR and netmasks that the SDN can allocate addressed from.  If this is specified, then DeprecatedClusterNetworkCIDR and DeprecatedHostSubnetLength may not be set.

pravisankar commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -24,29 +26,53 @@ func SetDefaultClusterNetwork(cn networkapi.ClusterNetwork) {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), clusterNet.HostSubnetLength, "subnet length is too large for clusterNetwork"))
-		} else if clusterNet.HostSubnetLength < 2 {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("hostSubnetLength"), clusterNet.HostSubnetLength, "subnet length must be at least 2"))
+	if len(clusterNet.Network) != 0 || clusterNet.HostSubnetLength != 0 {
```
Why we need to validate Network/HostSubnetLength if we only operate on ClusteNetworks? Don't we move deprecated values to ClusterNetwork[0] already

knobunc replied:
@danwinship wanted the validation, so @JacobTanenbaum added it back in.  I vote we leave it for now.

pravisankar commented:
@JacobTanenbaum Also needs corresponding docs PR

JacobTanenbaum commented:
@pravisankar I started one https://github.com/openshift/openshift-docs/pull/5382

JacobTanenbaum commented:
/test cmd

knobunc commented:
/lgtm

pravisankar commented:
/lgtm

smarterclayton commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -543,10 +543,12 @@ type RoutingConfig struct {
-	HostSubnetLength uint32 `json:"hostSubnetLength"`
+	// DeprecatedClusterNetworkCIDR is the CIDR string to specify the global overlay network's L3 space.  Deprecated, but maintained for backwards compatibility, use ClusterNetworks instead.
+	DeprecatedClusterNetworkCIDR string `json:"clusterNetworkCIDR,omitempty"`
+	// ClusterNetworks is a list of ClusterNetwork objects that defines the global overlay network's L3 space by specifying a set of CIDR and netmasks that the SDN can allocate addressed from.  If this is specified, then DeprecatedClusterNetworkCIDR and DeprecatedHostSubnetLength may not be set.
```
Use the external field name - a user won't have any idea what "DeprecatedClusterNetworkCIDR" is without looking at the go code.  In general, always use external field names in thees docs, no exceptions.

smarterclayton commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -543,10 +543,12 @@ type RoutingConfig struct {
-	HostSubnetLength uint32 `json:"hostSubnetLength"`
+	// ClusterNetworkCIDR is the CIDR string to specify the global overlay network's L3 space.  Deprecated, but maintained for backwards compatibility, use ClusterNetworks instead.
+	DeprecatedClusterNetworkCIDR string `json:"clusterNetworkCIDR,omitempty"`
+	// ClusterNetworks is a list of ClusterNetwork objects that defines the global overlay network's L3 space by specifying a set of CIDR and netmasks that the SDN can allocate addressed from.  If this is specified, then DeprecatedClusterNetworkCIDR and DeprecatedHostSubnetLength may not be set.
```
My comments got lost AGAIN.  Basically you can't use external names here.  Use the names as they would appear in JSON.  Users only see this from the API and documentation, they never see the go code, so the names in Go are meaningless.

`clusterNetworkCIDR` and `hostSubnetLength` are the names they see

knobunc commented:
@smarterclayton Ok, good to go?

JacobTanenbaum commented:
/test cmd

smarterclayton commented:
API approved

knobunc commented:
/lgtm

@smarterclayton would you approve this please too?  Thanks

eparis commented:
/approve

JacobTanenbaum commented:
flake #16414 /test end_to_end

eparis commented:
/retest
